eos/utils/kmatrix-impl.hh

```diff
@@ -0,0 +1,213 @@
+            }
+
+        ///////////////////
+        // 2. Fill Khat
+        ///////////////////
+        for (size_t i = 0 ; i < nchannels_ ; ++i)
+            {
+            for (size_t j = 0 ; j < nchannels_ ; ++j)
+                {
+                complex<double> entry = complex<double>(bkgcst[i][j].evaluate(), 0.0);
+
+                for (size_t a = 0 ; a < nresonances_ ; ++a)
+                    {
+
+                    double mres = resonances[a]->_m;
+
+                    Parameter g0rci = channels[i]->_g0s[a];
+                    Parameter g0rcj = channels[j]->_g0s[a];
+
+
+                    // Avoid the pole by adding a small imaginary part
+                    complex<double> mres2 = { mres * mres, 1.0e-7 };
+                    entry += g0rci * g0rcj / (mres2 - s);
+
+                    }
+
+                gsl_matrix_complex_set(_Khat, i, j, gsl_complex_rect(entry.real(), entry.imag()));
+                }
+            }
+
+        ///////////////////
+        // 3. Compute That
+        ///////////////////
+        static const gsl_complex one  = gsl_complex_rect(1.0, 0.0);
+        static const gsl_complex minusi  = gsl_complex_rect(0.0, -1.0);
+        static const gsl_complex zero = gsl_complex_rect(0.0, 0.0);
+
+        // 3a. multiply -i*rho with Khat from the right
+        // -> tmp_2 = -i * tmp_1 * Khat + zero * tmp_2 ; no transpose anywhere
+        gsl_blas_zgemm(CblasNoTrans, CblasNoTrans, minusi, _tmp_1, _Khat, zero, _tmp_2);
+
+        // 3b. add 1.0 to the diagonal elements of tmp_2
+        for (unsigned i = 0 ; i < nchannels_ ; ++i)
+            {
+            gsl_complex value = gsl_matrix_complex_get(_tmp_2, i, i);
+            gsl_matrix_complex_set(_tmp_2, i, i, gsl_complex_add_real(value, 1.0));
+            }
+
+        // 3c. invert tmp_2 and assign the result to tmp_1
+        int signum = 0;
+        gsl_permutation_init(_perm);
+        gsl_linalg_complex_LU_decomp(_tmp_2, _perm, &signum);
+        gsl_linalg_complex_LU_invert(_tmp_2, _perm, _tmp_1);
+
+        // 3d. calculate That = Khat / (1 - i rho * Khat) = Khat * tmp_1
+        gsl_blas_zgemm(CblasNoTrans, CblasNoTrans, one, _Khat, _tmp_1, zero, _That);
+
+
+        ///////////////////
+        // 4. Extract T matrix row
+        ///////////////////
+        for (size_t i = 0 ; i < nchannels_ ; ++i)
+        {
+            gsl_complex value = gsl_matrix_complex_get(_That, rowindex, i);
+            complex<double> cvalue(GSL_REAL(value), GSL_IMAG(value));
+            tmatrixrow[i] = cvalue;
+        }
+        return tmatrixrow;
+    }
+
+    template <unsigned nchannels_, unsigned nresonances_>
+    double
+    KMatrix<nchannels_, nresonances_>::partial_width(unsigned resonance, unsigned channel) const
+    {
+
+        double mres = this->_resonances[resonance]->_m;
+        double rho  = real(this->_channels[channel]->rho(mres*mres));
+
+        return pow(this->_channels[channel]->_g0s[resonance], 2.)/ mres * rho;
```

Comment:
- ```pow(..., 2.0)```: for integer powers we have ```power_of<unsigned>```, which uses multiplication for small powers, and can be inlined when many different powers are needed.
 - please leave a space to the left of ```/```.